src/hotspot/share/oops/klass.inline.hpp

```diff
@@ -76,4 +77,60 @@ inline ByteSize Klass::vtable_start_offset() {
   return in_ByteSize(InstanceKlass::header_size() * wordSize);
 }
 
+inline bool Klass::lookup_secondary_supers_table(Klass* k) const {
+  uintx bitmap = _bitmap;
+
+  if (bitmap == SECONDARY_SUPERS_BITMAP_FULL) {
+    return linear_search_secondary_supers(k);
+  }
+
+  constexpr int highest_bit_number = SECONDARY_SUPERS_TABLE_SIZE - 1;
+  uint8_t slot = k->_hash_slot;
+  uintx shifted_bitmap = bitmap << (highest_bit_number - slot);
+
+  // First check the bitmap to see if super_klass might be present. If
+  // the bit is zero, we are certain that super_klass is not one of
+  // the secondary supers.
+  if ((shifted_bitmap >> highest_bit_number) == 0) {
+    return false;
+  }
+
+  // Calculate the initial hash probe
+  int index = population_count(shifted_bitmap) - 1;
+  if (secondary_supers()->at(index) == k) {
+    // Yes! It worked the first time.
+    return true;
+  }
+
+  // Is there another entry to check? Consult the bitmap. If Bit 1,
+  // the next bit to test, is zero, we are certain that super_klass is
+  // not one of the secondary supers.
+  bitmap = rotate_right(bitmap, slot);
+  if ((bitmap & 2) == 0) {
+    return false;
+  }
+
+  // Continue probing the hash table
+  return fallback_search_secondary_supers(k, index, bitmap);
+}
+
+inline bool Klass::search_secondary_supers(Klass *k) const {
```

Comment:
I see you moved `Klass::search_secondary_supers` in `klass.inline.hpp`, but I'm not sure how it interacts with `Klass::is_subtype_of` (the sole caller) being declared in `klass.hpp`. 

Will the inlining still happen if `Klass::is_subtype_of()` callers include `klass.hpp`?

Reply:
Presumably this question applies to every function in `klass.inline.hpp`?
Practically everything does `#include "oops/klass.inline.hpp"`. It's inlined in about 120 files, as far as I can see everywhere such queries are made.

Reply:
My confusion arises from the following:
* `Klass::is_subtype_of()` is declared in `klass.hpp`
* `Klass::is_subtype_of()` calls `Klass::search_secondary_supers()`
* `Klass::search_secondary_supers()` is declared in `klass.inline.hpp`
* `klass.inline.hpp` includes `klass.hpp`

What happens when users include `klass.hpp`, but not `klass.inline.hpp`? How does it affect generated code? 

I suspect that `Klass::search_secondary_supers()` won't be inlinined in such case.

Reply:
> What happens when users include `klass.hpp`, but not `klass.inline.hpp`? How does it affect generated code?
> 
> I suspect that `Klass::search_secondary_supers()` won't be inlinined in such case.

That is true. I can't tell from this exchange whether you think it should. The same "wont be inlined" fact is also true of everything else in `klass.inline.hpp`.

Reply:
> I suspect that Klass::search_secondary_supers() won't be inlinined in such case.

That's true, but it's true of every other function in that file. Is it not deliberate?

Reply:
FYI, somewhat related: AArch64 GCC inlines `lookup_secondary_supers_table()` 237 times (it's only a few instructions.)
x86-64 GCC, because it doesn't use a popcount intrinsic, decides that `lookup_secondary_supers_table()` is too big to be worth inlining in all but 3 cases. So the right thing happens, I think: where we can profit from fast lookups without bloating the runtime, we do.

Reply:
> That's true, but it's true of every other function in that file. Is it not deliberate?

IMO the fact that `Klass::search_secondary_supers()` is used in `klass.hpp` makes a difference here. 

After thinking more about it, I did a small experiment [1] and observed a build failure on AArch64 [2]. 
I think we don't see any more failures simply because `klass.inline.hpp` is included pervasively.

What do you think about moving `Klass::is_subtype_of()` to `klass.inline.hpp`? 

[1]
```
diff --git a/test/hotspot/gtest/oops/test_klass.cpp b/test/hotspot/gtest/oops/test_klass.cpp
new file mode 100644
index 00000000000..326a70f1f54
--- /dev/null
+++ b/test/hotspot/gtest/oops/test_klass.cpp
@@ -0,0 +1,9 @@
+#include "precompiled.hpp"
+#include "oops/klass.hpp"
+#include "unittest.hpp"
+
+TEST_VM(Klass, is_subtype_of) {
+  Klass* k = vmClasses::Object_klass();
+  ASSERT_TRUE(k->is_subtype_of(k));
+}
```

[2] 
```
Undefined symbols for architecture arm64:
  "Klass::search_secondary_supers(Klass*) const", referenced from:
      Klass_is_subtype_of_vm_Test::TestBody() in test_klass.o
```